File: healthcare-backend/src/main/java/com/healthcare/service/NotificationService.java
package com.healthcare.service;

import com.healthcare.entity.AuthorizationRequest;
import com.healthcare.entity.Notification;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;

import java.util.List;

public interface NotificationService {

    /**
     * Creates notifications for both Provider and Payer when a status changes.
     */
    void notifyStatusChange(AuthorizationRequest request, AuthorizationRequest.RequestStatus newStatus);

    /**
     * Retrieves all notifications for a given recipient (paginated).
     */
    Page<Notification> getNotifications(String recipientId, Pageable pageable);

    /**
     * Retrieves only unread notifications for a given recipient.
     */
    List<Notification> getUnreadNotifications(String recipientId);

    /**
     * Returns the count of unread notifications for a recipient.
     */
    long getUnreadCount(String recipientId);

    /**
     * Marks a single notification as read.
     */
    Notification markAsRead(String notificationId);

    /**
     * Marks all notifications for a recipient as read.
     */
    void markAllAsRead(String recipientId);
}

File: healthcare-frontend/src/app/services/notification.service.ts
import { Injectable } from '@angular/core';
import { HttpClient, HttpParams } from '@angular/common/http';
import { Observable } from 'rxjs';
import { environment } from '../../environments/environment';
import { Page } from './authorization.service';

export interface Notification {
  id: string;
  recipientId: string;
  authorizationRequestId: string;
  title: string;
  message: string;
  type: string;
  read: boolean;
  createdAt: string;
}

@Injectable({
  providedIn: 'root'
})
export class NotificationService {
  private apiUrl = `${environment.apiUrl}/notifications`;

  constructor(private http: HttpClient) {}

  getNotifications(recipientId: string, page = 0, size = 10): Observable<Page<Notification>> {
    const params = new HttpParams().set('page', page).set('size', size);
    return this.http.get<Page<Notification>>(`${this.apiUrl}/${recipientId}`, { params });
  }

  getUnreadNotifications(recipientId: string): Observable<Notification[]> {
    return this.http.get<Notification[]>(`${this.apiUrl}/${recipientId}/unread`);
  }

  getUnreadCount(recipientId: string): Observable<{ unreadCount: number }> {
    return this.http.get<{ unreadCount: number }>(`${this.apiUrl}/${recipientId}/unread/count`);
  }

  markAsRead(notificationId: string): Observable<Notification> {
    return this.http.patch<Notification>(`${this.apiUrl}/${notificationId}/read`, null);
  }

  markAllAsRead(recipientId: string): Observable<void> {
    return this.http.patch<void>(`${this.apiUrl}/${recipientId}/read-all`, null);
  }
}
